x/programs/runtime/dependencies.go

```diff
@@ -30,7 +30,7 @@ type Runtime interface {
 	// module with the given imports.  Initialize should only be called once.
 	Initialize(context.Context, []byte, uint64) error
 	// Call invokes the an exported guest function with the given parameters.
-	Call(context.Context, string, ...uint64) ([]uint64, error)
+	Call(context.Context, string, ...int64) ([]int64, error)
```

Comment:
While I see your point technically, I would like to see a strong usecase to make this change. Wazero[1] made the same design choice and I feel it is the right one for the sake of simplicity. Now one could argue that this could be an interface{} but again the amount of complexity we save by assuming unsigned ints is non trivial.

[1] https://github.com/tetratelabs/wazero/blob/d3b83a596cde165d0fe69bc4992f31801e516be3/api/wasm.go#L386

Question: what would the usecase be for an exported function to take a param that has a negative value? Could this be solved in other ways with a tradeoff being simplicity?

Reply:
Hmm, why do you say assuming unsigned ints decreases complexity? The wasmtime `call` [function](https://pkg.go.dev/github.com/bytecodealliance/wasmtime-go#Func.Call) converts any `uint` to `int` which can potentially overflow. Therefore we get no benefit from using `uint64`(since anything above `Int64Max` will cause overflow) _AND_ we limit our programs to only unsigned numbers. 

So we could use an interface{} and accept `int32`, `int64` and `uint32` but as you said this would increase complexity. And I don't think `interface{}` is worth it here.

There are plenty of use cases for signed ints. The `rust sdk` is a tool for building programs, so any use case for signed ints in traditional programs are use cases for signed ints in wasm programs. Of course any function that would like to accept a negative number could do something like `reduceBalanceBy(uint balance)` & `increaseBalance(uint balance)` but I feel like this is an artificial workaround that increases complexity on the program dev.

Reply:
Yeah since we really won't be supporting floats I think interface{} is also overkill. While I think that programs although arbitrary by definition will be very much tied to transactions. So in the context of a VM it seems unsigned ints are really a much more native type and IMO easier to deal with.

But I do see that it is used in for example Solidity to represent debts as you outlined above. So I don't want to make the interface overly opinionated. Lets think about this a little bit more but I am leaning towards the change. Thanks for the feedback.

Reply:
Yeah I think this makes sense for a lot of reasons. In the future we will probably be passing byte slices to int64 types exclusively. But I think this change makes sense for now.